### PR TITLE
test: Backed enum resource tests

### PR DESCRIPTION
**Pull request by GwendolenLynch**
| Q             | A
| ------------- | ---
| Branch?       | 3.2
| Tickets       | #6264 #6279 #6283 
| License       | MIT

@soyuka this is just an updated version of the tests I linked yesterday when discussing #6264 & #6279 … and if I m still wrong please just close this PR. :+1: 

These tests illustrate what *was* working for `json`, `jsonhal`, and `jsonld` formats, but breaks now with the change in https://github.com/symfony/symfony/pull/54315

However, if a fix/workaround is found these tests will still add additional coverage for enum resources.

**Notes:**
- Tests pass when with:
  -  core `3.2` as-of `v3.2.19` (i.e. pre #6283)
  - `symfony/framework-bundle` `<=7.0.5` (i.e. pre https://github.com/symfony/symfony/pull/54315)
- ~Missing `jsonapi` tests due to target branch~ (Derp! serializer !== schema factory)

### CODE REVIEW

**Comment by GwendolenLynch** [comment]
Symfony have merged [#54484](https://github.com/symfony/symfony/pull/54484) so I've taken this out of draft.

I've added a conflict for `6.4.6 || 7.0.6` of `symfony/framework-bundle` and that should have core covered.

File: tests/Functional/BackedEnumResourceTest.php [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+        $this->assertResponseIsSuccessful();
+        $this->assertJsonEquals([
+            '@context' => '/contexts/Availability',
+            '@id' => '/availabilities',
+            '@type' => 'hydra:Collection',
+            'hydra:totalItems' => 3,
+            'hydra:member' => [
+                [
+                    '@id' => '/availabilities/AVAILABLE',
```
Comment (soyuka) [inline tests/Functional/BackedEnumResourceTest.php:183]: this is wrong on an RDF point of view, and as I pointed I have an ongoing work but can't merge/work on that right now

**Review by soyuka — COMMENTED** [review]
I really like this but I won't be able to merge it right away as we need to work on the Hydra implementation. I reverted the priority change on our source code https://github.com/api-platform/core/commit/2a8767108690c35d873a936f9e2383365fdb4f00

**Comment by soyuka** [comment]
also needs https://github.com/soyuka/core/commit/48db6fa3e63f2207048069db99c022898860f5e1

**Comment by soyuka** [comment]
Nice addition, API Platform supports enums only on properties for documentation via OpenAPI and using GraphQl but we did not add the support for RDF formats such as JSON-LD. 
I started working on this (https://github.com/soyuka/core/commit/48db6fa3e63f2207048069db99c022898860f5e1) but it needs more work. Currently it implements the context:


```php
#[ApiResource(operations: [])] // An enum can't have operations so this needs to be fixed in he Metadata Factory
enum BackgroundColor: string
{
    case Sand = 'sand';
    case Magenta = 'magenta';
    case Teal = 'teal';
    case Blue = 'blue';
    case Mauve = 'mauve';

    /**
     * @return string[]
     */
    public static function values(): array
    {
        return array_column(self::cases(), 'value');
    }
}
```

enumcontext

My reference for the work is what's done on schema.org, for example:

https://schema.org/EnergyEfficiencyEnumeration 

This is of type [Enumeration](https://schema.org/Enumeration) and you can see it's "Enumeration members" at https://schema.org/EUEnergyEfficiencyEnumeration. When looking at the schema.org jsonld https://github.com/schemaorg/schemaorg/blob/main/data/releases/26.0/schemaorg-all-https.jsonld we find lots of examples:

```json
{
      "@id": "schema:GamePlayMode",
      "@type": "rdfs:Class",
      "rdfs:comment": "Indicates whether this game is multi-player, co-op or single-player.",
      "rdfs:label": "GamePlayMode",
      "rdfs:subClassOf": {
        "@id": "schema:Enumeration"
      }
},
```

https://github.com/schemaorg/schemaorg/blob/c7dc8e5a3ad9f1e1645e282a18e46b5d2c19a657/data/releases/26.0/schemaorg-all-https.jsonld#L19759-L19767

And one of the value of the enum has this type:

```json
{
      "@id": "schema:SinglePlayer",
      "@type": "schema:GamePlayMode",
      "rdfs:comment": "Play mode: SinglePlayer. Which is played by a lone player.",
      "rdfs:label": "SinglePlayer"
},
```

https://github.com/schemaorg/schemaorg/blob/c7dc8e5a3ad9f1e1645e282a18e46b5d2c19a657/data/releases/26.0/schemaorg-all-https.jsonld#L24689-L24694 (there are other values on this enum, just like for EnergyEfficiency).

So, for each value of the Backed Enum we should provide a given type to be referenced in the `docs.jsonld`.

**Comment by GwendolenLynch** [comment]
> I reverted the priority change on our source code https://github.com/api-platform/core/commit/2a8767108690c35d873a936f9e2383365fdb4f00

Suggestion on that, it is probably safe to revert the change to `docs.feature` as well.

Oh, and it is probably worth at least cherry-picking the composer change from here as that will prevent Symfony "broken" versions.

**Comment by GwendolenLynch** [comment]
> I really like this but I won't be able to merge it right away as we need to work on the Hydra implementation.

No stress. I was more trying to get coverage for how it is now, as it is "in the wild" as far as use (I'm also guilty there) … supported or not.

**Comment by soyuka** [comment]
> Suggestion on that, it is probably safe to revert the change to docs.feature as well.

Actually this is a bad test, I already picked the change on `main` (it's due to a property-info fix) and has no impact on the expected feature. 

> Suggestion on that, it is probably safe to revert the change to docs.feature as well.

I plan to do that indeed.

> No stress. I was more trying to get coverage for how it is now, as it is "in the wild" as far as use (I'm also guilty there) … supported or not.

No problem we can work on integrating this, I'm afraid that it won't get into 3.3 as I'm already a month late on that release but it's definitely going to be a top priority for API Platform 3.4. 
Anyways if you want to work on this feel free to do so, I'll see if I can in the upcoming weeks although I'll be afk for 3 weeks (except on weekends). I'll open a new PR with this test and my work for starters.

File: tests/Functional/BackedEnumResourceTest.php [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                    '@type' => 'Availability',
+                    'value' => 20,
+                ],
+                [
+                    '@id' => '/availabilities/POSTPONED',
+                    '@type' => 'Availability',
+                    'value' => 30,
+                ],
+            ],
+        ]);
```
Comment (soyuka) [inline tests/Functional/BackedEnumResourceTest.php:198]: this is probably correct, @dunglas could you review these tests?

**Comment by GwendolenLynch** [comment]
To catch future versions of some of the issues hit, I have added test for `int` and `string` enum resources, as well as GraphQL tests too. 

The latter seems to have uncovered some internal-use of deprecated code in the GraphQL component.

File: composer.json [hunk 3]
```diff
@@ -82,17 +82,18 @@
         "symfony/expression-language": "^6.1 || ^7.0",
         "symfony/finder": "^6.1 || ^7.0",
         "symfony/form": "^6.1 || ^7.0",
-        "symfony/framework-bundle": "^6.1 || ^7.0",
+        "symfony/framework-bundle": "^6.4.3 || ^7.0",
```
Comment (soyuka) [inline composer.json:85]: let's target main or remove composer changes?

Comment (GwendolenLynch) [reply composer.json:86]: Updated to `main`. There is no way to get `--prefer-lowest` to pass without these.

Comment (soyuka) [reply composer.json:86]: but is this tight to enums?

Comment (GwendolenLynch) [reply composer.json]: `symfony/framework-bundle` change seems to be not needed now the rest of the SF components are `^6.4` so I've dropped it.

`symfony/phpunit-bridge` is to include https://github.com/symfony/symfony/issues/52844 which meant the tests just wouldn't run
```shell
Warning: The lock file is not up to date with the latest changes in composer.json. You may 
be getting outdated dependencies. It is recommended that you run `composer update` or 
`composer update <package name>`.
- Required package "symfony/phpunit-bridge" is not present in the lock file.
```

`symfony/string` is 100% about enums, and was the hardest to track down at the time. The change does mirror the constraint used in the component `composer.json` files.

Comment (GwendolenLynch) [reply composer.json]: Anticipating the question "Why isn't CI failing?" (I had that question too), see this example from `main` yesterday:
https://github.com/api-platform/core/actions/runs/8679784849/job/23799224716?pr=6309
> PHPUnit is not installed. Please run ./vendor/bin/simple-phpunit to install it

There seems to be a new-ish issue with CI set-up.

Originally I caught the failure on CI and then moved to local testing where it work correctly with `6.4.1`+

Comment (GwendolenLynch) [reply composer.json]: See #6113

File: tests/Functional/BackedEnumResourceTest.php (hunk 2 above)
Comment (dunglas) [reply tests/Functional/BackedEnumResourceTest.php:259]: LGTM

File: composer.json [hunk 4]
```diff
@@ -89,11 +89,12 @@
         "symfony/maker-bundle": "^1.24",
         "symfony/mercure-bundle": "*",
         "symfony/messenger": "^6.4 || ^7.0",
-        "symfony/phpunit-bridge": "^6.4 || ^7.0",
+        "symfony/phpunit-bridge": "^6.4.1 || ^7.0",
```
Comment (soyuka) [inline composer.json:92]:
```suggestion
```

Comment (GwendolenLynch) [reply composer.json:92]: It will disappear on rebase when added to #6316 :+1:

File: tests/Functional/BackedEnumResourceTest.php [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        $this->assertResponseIsSuccessful();
+        $this->assertJsonEquals([
+            '@context' => '/contexts/Availability',
+            '@id' => '/availabilities',
+            '@type' => 'hydra:Collection',
+            'hydra:totalItems' => 3,
+            'hydra:member' => [
+                [
+                    '@id' => '/availabilities/0',
+                    '@type' => 'Availability',
```
Comment (soyuka) [inline tests/Functional/BackedEnumResourceTest.php:245]:
```suggestion
                    '@type' => 'Available',
```

File: tests/Functional/BackedEnumResourceTest.php [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+                    'value' => 0,
+                ],
+                [
+                    '@id' => '/availabilities/10',
+                    '@type' => 'Availability',
```
Comment (soyuka) [inline tests/Functional/BackedEnumResourceTest.php:250]:
```suggestion
                    '@type' => 'Cancelled',
```

File: tests/Functional/BackedEnumResourceTest.php [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+                    'value' => 10,
+                ],
+                [
+                    '@id' => '/availabilities/200',
+                    '@type' => 'Availability',
```
Comment (soyuka) [inline tests/Functional/BackedEnumResourceTest.php:255]:
```suggestion
                    '@type' => 'Postponed',
```

**Comment by soyuka** [comment]
also we need to add the enum to the jsonld context, something like this: 

```
    {
      "@id": "schema:CoOp",
      "@type": "schema:GamePlayMode",
      "rdfs:comment": "Play mode: CoOp. Co-operative games, where you play on the same team with friends.",
      "rdfs:label": "CoOp"
    },
{
      "@id": "schema:GamePlayMode",
      "@type": "rdfs:Class",
      "rdfs:comment": "Indicates whether this game is multi-player, co-op or single-player.",
      "rdfs:label": "GamePlayMode",
      "rdfs:subClassOf": {
        "@id": "schema:Enumeration"
      }
    },
    {
      "@id": "schema:SinglePlayer",
      "@type": "schema:GamePlayMode",
      "rdfs:comment": "Play mode: SinglePlayer. Which is played by a lone player.",
      "rdfs:label": "SinglePlayer"
    },
```

**Comment by soyuka** [comment]
tyvm @GwendolenLynch ! Looking forward to merging the enum feature next (I still need to try the branch and see what I can improve on the URI thing).

**Pull request merged**